RFCs/0016-Loggers.md

```diff
@@ -0,0 +1,85 @@
+# 0016 Loggers
+
+## Summary
+This note outlines the proposed changes for:
+1. Enabling logger support in test platform for TranslationLayer clients.
+2. Allowing users to provide loggers from the runsettings.
+
+## Motivation
+1. Enabling logger support for TranslationLayer clients will allow clients to use logger extensibility feature.
+
+2. Allowing users to provide loggers from the runsettings enables:
+  * TranslationLayer clients to provide loggers to the test platform.
+  * Alternative way to provide loggers in addition to the existing command line logger argument.
+
+## Specifying a logger
+Loggers can be provided to the test platform in one of the following ways:
+
+1.  "/Logger:"UriOrFriendlyName";Key1=Value1;Key2=Value2" This is a switch to vstest.console.exe that feeds logger information to the test platform. For instance:
+
+    ```
+    /Logger:"logger://Microsoft/TestPlatform/TrxLogger/v1"
+    ```
+
+    This loads and initializes logger with "logger://Microsoft/TestPlatform/TrxLogger/v1" as uri.
+
+    ```
+    /Logger:sampleLogger;Key1=Value1;Key1=Value2
+    ```
+
+    This loads and initializes logger with "sampleLogger" as friendly name. Key1=Value1 and Key2=Value2 are passed as dictionary parameters to the logger while initialization.
+
+2. Runsettings via "Logger" node in the LoggerRunSettings section. Here is a sample on how this can be specified:
+
+    ```xml
+    <RunSettings>
+        <LoggerRunSettings>
+            <Loggers>
+                <Logger friendlyName="sampleLoggerwithParameters">
```

Comment:
What are the errors user will see if a logger is incorrectly specified? Will the test run continue or it will be aborted?

Reply:
Exceptions are thrown in following scenarios:
1. `LoggerRunSettings` or `Loggers` node has attributes.
2. `LoggerRunSettings` has any node other than `Loggers`.
3. `Loggers` has any node other than `Logger`.
4. `Logger` has any node other than `Configuration`.
5. `Logger` node has any attribute other than `uri`, `assemblyQualifiedName`, `friendlyName`, or `enabled`.
6. Invalid format uri is given.
7. Unable to find logger using precedent attribute.
8. If none of the attributes `uri`, `assemblyName`, `friendlyName` is given.

Warning are printed in following scenarios:
1. Key value pair has empty or whitespace value. Example: `<Key1></Key1>`.

Duplicate keys in configuration is ignored and verbose is added for it.

I will add these details in RFC as well.
In addition, `Logger` node understands enabled attribute as well. I will update RFC regarding the same.

Reply:
Assuming exceptions will abort the test run.

Reply:
Yes, exceptions will abort the test run.

Reply:
What I recall is that IDE also tries to read the runsettings. If something is wrong with settings, IDE aborts the run and request does not reach the TestPlatform.
Also make sure we catch these exceptions, give out an actionable message to user and dont crash vstest.console exe.

---
In reply to: [160580427](https://github.com/Microsoft/vstest-docs/pull/111#discussion_r160580427) [](ancestors = 160580427)

Reply:
Sure. I will ensure, that vstest.console does not crash and only test aborts.